profiler/src/main/java/com/splunk/opentelemetry/profiler/allocation/exporter/PprofAllocationEventExporter.java

```diff
@@ -0,0 +1,291 @@
+                method.getType().getName() + "." + method.getName(),
+                frame.getLineNumber()));
+      }
+    }
+    if (spanContext != null && spanContext.isValid()) {
+      Label.Builder traceIdLabel = Label.newBuilder();
+      traceIdLabel.setKey(stringTable.get("trace_id"));
+      traceIdLabel.setStr(stringTable.get(spanContext.getTraceId()));
+      sample.addLabel(traceIdLabel.build());
+
+      Label.Builder spanIdLabel = Label.newBuilder();
+      spanIdLabel.setKey(stringTable.get("span_id"));
+      spanIdLabel.setStr(stringTable.get(spanContext.getSpanId()));
+      sample.addLabel(spanIdLabel.build());
+
+      Label.Builder eventTypeLabel = Label.newBuilder();
+      eventTypeLabel.setKey(stringTable.get("event_type"));
+      eventTypeLabel.setStr(stringTable.get(event.getEventType().getName()));
+      sample.addLabel(eventTypeLabel.build());
+    }
+    // XXX are there any more attributes that should be added to the sample?
+    // XXX PlainTextAllocationEventExporter adds attributes from the used AllocationEventSampler
+
+    profile.addSample(sample);
+  }
+
+  private byte[] serializePprof() {
+    Profile profile = profileData.profile.build();
+    profileData = new ProfileData();
+
+    try {
+      ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+      boolean gzip =
+          allocationDataFormat == AllocationDataFormat.PPROF_GZIP
+              || allocationDataFormat == AllocationDataFormat.PPROF_GZIP_BASE64;
+      try (OutputStream outputStream = gzip ? new GZIPOutputStream(byteStream) : byteStream) {
+        profile.writeTo(outputStream);
+      }
+      byte[] bytes = byteStream.toByteArray();
+      if (allocationDataFormat == AllocationDataFormat.PPROF_BASE64
+          || allocationDataFormat == AllocationDataFormat.PPROF_GZIP_BASE64) {
+        bytes = Base64.getEncoder().encode(bytes);
+      }
+      return bytes;
+    } catch (IOException exception) {
+      throw new IllegalStateException("Failed to serialize pprof", exception);
+    }
+  }
+
+  public static final AttributeKey<String> DATA_TYPE = stringKey("com.splunk.datatype");
+  public static final AttributeKey<String> ENCODING = stringKey("com.splunk.encoding");
+
+  @Override
+  public void flush() {
+    // Flush is called after each JFR chunk, hopefully this will keep batch sizes small enough.
+    byte[] bytes = serializePprof();
+    String encoding = allocationDataFormat.toString().toLowerCase(Locale.ROOT).replace('_', '-');
+
+    // XXX just to give an overview of exported data size
+    logger.info("Exporting {}, size {}", encoding, bytes.length);
+
+    Attributes attributes =
+        Attributes.builder()
+            .put(SOURCE_TYPE, PROFILING_SOURCE)
+            .put(DATA_TYPE, "allocation") // tells that this message is about allocation samples
+            .put(ENCODING, encoding) // data encoding
+            .build();
+
+    String body = new String(bytes, StandardCharsets.ISO_8859_1);
```

Comment:
Hmm, is this charset a part of some spec? Also, this is a lot of allocation, I wonder if we can somehow avoid this, like does the OTLP format mandate it as a string?

Reply:
Currently api only allows adding string as body.

Reply:
Aha, got it, a possible hackaround. `ISO_8859_1` still has a bunch of undefined values, and the data might contain eight 0 bits in a row, I would expect something, somewhere in the pipeline, to break because of it. Will try

Reply:
So apparently this does not work -- we have to do base64. Otherwise bytes like `0x00, 0x07, 0x0F, 0x10, 0x17, 0x1F, 0x7F, 0x80, 0x87, 0x8F, 0x90, 0x97, 0x9F, 0xD7, 0xF7` will be corrupted on the way, both collector and ingest will change them during JSON conversion.

Resolve.